DEVELOPERS_GUIDE.md

```diff
@@ -0,0 +1,87 @@
+
+ 
+### Data Source 
+
+DataSource is an interface that allows one to define specializations for a source of GC events that maybe of interest. Currently the toolkit contains two data sources, a FileDataSource and a SafepointogFile. A GClogFile is a specialization of a FileDataSource. This is in turn has two specializations, SingleGCLogFIle and RotatingGCLogFile. The concrete classes take care of the details of being able to create a continuous stream of their contents.
+
+Another role of the DataSource is to provide a Diary (or summary) of important features found in the data. This would include information such as, is this a unified log or a pre-unified log, which version of the JVM produced the file. What flags where used to produce the file, which collector (combinations) are in play and so on. This information is used to construct the data paths through GCToolkit. It may also be used by clients of GCToolkit for their own purposes. For example, analytics may use the information to modulate how they process data.
+
+An example of this would be an analytic that looked at System.gc() behaviour. In JDK 1.7.0_40 the logs stopped reporting user triggered collections unless a new flag was enabled. This left the analytic blind to these events. In JDK 1.8.0, this new flag was automatically turned on when the PrintGCDetails flag was set. Making the analytics aware of the JDK version allowed them to calculate when System.gc() was likely being called thus enabling the analysis when the cause was missing in the log.
+
+###Parsers
+
+The parsers extract information from the log files and generation JVMEvents. All JVMEvents share the same 3 characteristics; the type of the event (reflected in the class), the time of the event, and the duration of the event. These properties are immutable and as such, the creation of a JVMEvent is delayed until they are known. The parse rules make heavy use of regular expressions. This is done so that the parser can make a best effort to extract information from a log even if that log is mixed with output from an unexpected source of the log or, the logs lines are some how corrupted.
+
+The currently existing parsers are generational, G1, Z, and Shenandoah. The generational parser cover the serial, parallel (PSYoung, PSOld), ParNew, CMS, and iCMS collectors. There are separate parsers for unified and pre-unified logs produced by the generational and G1 collectors. The pre-unified generational collectors should cope with logs produced by JDK 1.4.2 whereas the G1 log parsers are for JDK 8+. ZGC support starts at JDK 16 whereas Shenandoah starts at JDK 11.
```

Comment:
Just a nit, but 'whereas' is used to compare and contrast - "The banker's hands were smooth and soft whereas the farmer's hands were hard and calloused". Since 'whereas' is a conjunction, you can fix this (if you care) by just removing and making separate sentences.